# CSV_READER

In [1]:
import pandas as pd

In [ ]:
url = 'https://raw.githubusercontent.com/alura-cursos/Pandas/refs/heads/main/superstore_data.csv'
dados_mercado= pd.read_csv(url , sep=',')
dados_mercado.head()

Id  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced  84835.0        0         0   
1      1        1961  Graduation         Single  57091.0        0         0   
2  10476        1958  Graduation        Married  67267.0        0         1   
3   1386        1967  Graduation       Together  32474.0        1         1   
4   5371        1989  Graduation         Single  21474.0        1         0   

  Dt_Customer  Recency  MntWines  ...  MntFishProducts  MntSweetProducts  \
0   6/16/2014        0       189  ...              111               189   
1   6/15/2014        0       464  ...                7                 0   
2   5/13/2014        0       134  ...               15                 2   
3   11/5/2014        0        10  ...                0                 0   
4    8/4/2014        0         6  ...               11                 0   

   MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
0           218                  1                4                    4   
1            37                  1                7                    3   
2            30                  1                3                    2   
3             0                  1                1                    0   
4            34                  2                3                    1   

   NumStorePurchases  NumWebVisitsMonth  Response  Complain  
0                  6                  1         1         0  
1                  7                  5         1         0  
2                  5                  2         0         0  
3                  2                  7         0         0  
4                  2                  7         1         0  

[5 rows x 22 columns]

In [ ]:
dado_selecao = pd.read_csv(url, usecols=['Id', 'Year_Birth', 'Income'])
dado_selecao

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

Também funciona indices

In [ ]:
dado_selecao = pd.read_csv(url, usecols=[0, 1, 4])
dado_selecao

Id  Year_Birth   Income
0      1826        1970  84835.0
1         1        1961  57091.0
2     10476        1958  67267.0
3      1386        1967  32474.0
4      5371        1989  21474.0
...     ...         ...      ...
2235  10142        1976  66476.0
2236   5263        1977  31056.0
2237     22        1976  46310.0
2238    528        1978  65819.0
2239   4070        1969  94871.0

[2240 rows x 3 columns]

In [ ]:
dado_selecao.to_csv('clientes_mercado.csv', index=False)

In [ ]:
clients_mercado= pd.read_csv('/content/clientes_mercado.csv')

#Planilhas Reader


In [ ]:
url_sheets = 'https://github.com/alura-cursos/Pandas/raw/refs/heads/main/emissoes_CO2.xlsx?=True'

In [ ]:
dados_co2= pd.read_excel(url_sheets)

In [ ]:
dados_co2.head()

País ISO 3166-1 alpha-3   Ano  Total  Carvão  Oléo  Gás  Cimento  \
0  Afeganistão                AFG  1750    0.0     NaN   NaN  NaN      NaN   
1  Afeganistão                AFG  1751    0.0     NaN   NaN  NaN      NaN   
2  Afeganistão                AFG  1752    0.0     NaN   NaN  NaN      NaN   
3  Afeganistão                AFG  1753    0.0     NaN   NaN  NaN      NaN   
4  Afeganistão                AFG  1754    0.0     NaN   NaN  NaN      NaN   

   Queima  Outro  Per Capita  
0     NaN    NaN         NaN  
1     NaN    NaN         NaN  
2     NaN    NaN         NaN  
3     NaN    NaN         NaN  
4     NaN    NaN         NaN

In [ ]:
pd.ExcelFile(url_sheets).sheet_names

['emissoes_C02', 'emissoes_percapita', 'fontes']

In [ ]:
percapita =pd.read_excel(url_sheets,sheet_name='emissoes_percapita')

In [ ]:
fontes = pd.read_excel(url_sheets,sheet_name='fontes')

usecols

In [ ]:
intervalo = pd.read_excel(url_sheets, usecols='A:D')

In [ ]:
intervalo.head()

País ISO 3166-1 alpha-3   Ano  Total
0  Afeganistão                AFG  1750    0.0
1  Afeganistão                AFG  1751    0.0
2  Afeganistão                AFG  1752    0.0
3  Afeganistão                AFG  1753    0.0
4  Afeganistão                AFG  1754    0.0

In [ ]:
intervalo2 = pd.read_excel(url_sheets, usecols='A:D',nrows=10)

In [ ]:
intervalo2

País ISO 3166-1 alpha-3   Ano  Total
0  Afeganistão                AFG  1750      0
1  Afeganistão                AFG  1751      0
2  Afeganistão                AFG  1752      0
3  Afeganistão                AFG  1753      0
4  Afeganistão                AFG  1754      0
5  Afeganistão                AFG  1755      0
6  Afeganistão                AFG  1756      0
7  Afeganistão                AFG  1757      0
8  Afeganistão                AFG  1758      0
9  Afeganistão                AFG  1759      0

In [ ]:
percapita.to_excel('co2_percapita.xlsx',index=False)

From google sheets

In [ ]:
https://docs.google.com/spreadsheets/d/12Oy0VxoI0gfb2ejkc84boR_Rjo9m-p1pLWJYf0UiIVw/edit?usp=sharing

In [ ]:
sheet_id = '12Oy0VxoI0gfb2ejkc84boR_Rjo9m-p1pLWJYf0UiIVw'

In [ ]:
url_google = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet'

In [ ]:
dados_co2_google= pd.read_csv(url_google)
dados_co2.head()

País ISO 3166-1 alpha-3   Ano  Total  Carvão  Oléo  Gás  Cimento  \
0  Afeganistão                AFG  1750    0.0     NaN   NaN  NaN      NaN   
1  Afeganistão                AFG  1751    0.0     NaN   NaN  NaN      NaN   
2  Afeganistão                AFG  1752    0.0     NaN   NaN  NaN      NaN   
3  Afeganistão                AFG  1753    0.0     NaN   NaN  NaN      NaN   
4  Afeganistão                AFG  1754    0.0     NaN   NaN  NaN      NaN   

   Queima  Outro  Per Capita  
0     NaN    NaN         NaN  
1     NaN    NaN         NaN  
2     NaN    NaN         NaN  
3     NaN    NaN         NaN  
4     NaN    NaN         NaN

In [ ]:
sheet_id = '12Oy0VxoI0gfb2ejkc84boR_Rjo9m-p1pLWJYf0UiIVw'
sheet_name = 'emissoes_percapita'
url_pagina = f'https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}'
dados_emissoes_percapito = pd.read_csv(url_pagina)
dados_emissoes_percapito.head()

País ISO 3166-1 alpha-3   Ano Total Carvão Óleo  Gás Cimento Queima  \
0  Afeganistão                AFG  1750   NaN    NaN  NaN  NaN     NaN    NaN   
1  Afeganistão                AFG  1751   NaN    NaN  NaN  NaN     NaN    NaN   
2  Afeganistão                AFG  1752   NaN    NaN  NaN  NaN     NaN    NaN   
3  Afeganistão                AFG  1753   NaN    NaN  NaN  NaN     NaN    NaN   
4  Afeganistão                AFG  1754   NaN    NaN  NaN  NaN     NaN    NaN   

   Outro  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN

#JSON_FILES

In [ ]:
import pandas as pd

In [ ]:
import json

In [ ]:
with open('/content/pacientes_2.json','r') as file :
  dados_pacientes = json.loads(file.read())

In [ ]:
dados_pacientes

{'Pesquisa': 'Principais Indicadores de Doenca Cardiaca',
 'Ano': 2020,
 'Pacientes': [{'ID': '01',
   'Faixa_etaria': '55-59',
   'Sexo_biologico': 'Feminino',
   'Raça': 'Branca',
   'IMC': 16.6,
   'Fumante': 'Sim',
   'Consumo_alcool': 'Nao',
   'Saude_fisica': 3,
   'Saude_mental': 30,
   'Dificuldade_caminhar': 'Nao',
   'Atividade_fisica': 'Sim',
   'Saude_geral': 'Muito boa',
   'Horas_sono': 5,
   'Problemas_saude': ['Diabetes', 'Asma', 'Cancer_pele']},
  {'ID': '02',
   'Faixa_etaria': '80 ou +',
   'Sexo_biologico': 'Feminino',
   'Raça': 'Branca',
   'IMC': 20.34,
   'Fumante': 'Nao',
   'Consumo_alcool': 'Nao',
   'Saude_fisica': 0,
   'Saude_mental': 0,
   'Dificuldade_caminhar': 'Nao',
   'Atividade_fisica': 'Sim',
   'Saude_geral': 'Muito boa',
   'Horas_sono': 7,
   'Problemas_saude': ['AVC']},
  {'ID': '03',
   'Faixa_etaria': '65-69',
   'Sexo_biologico': 'Masculino',
   'Raça': 'Branca',
   'IMC': 26.58,
   'Fumante': 'Sim',
   'Consumo_alcool': 'Nao',
   'Saude_fis

olhando ai acima ja sabemos que o paciente vai ta aninhado e precisamos concertar ele e nao so isso temos que aparecer o json de pesquisa e de ano. Por isso preisamos adicionar esse meta, e o record_path é a coisa que queremos desaninhar

In [ ]:
df_normalizado = pd.json_normalize(dados_pacientes,record_path='Pacientes',meta=['Pesquisa','Ano'])

In [ ]:
df_normalizado.to_json('Historico_pacientes_normalizados.json')

In [ ]:
pd.read_json('/content/Historico_pacientes_normalizados.json')

ID Faixa_etaria Sexo_biologico    Raça    IMC Fumante Consumo_alcool  \
0   1        55-59       Feminino  Branca  16.60     Sim            Nao   
1   2      80 ou +       Feminino  Branca  20.34     Nao            Nao   
2   3        65-69      Masculino  Branca  26.58     Sim            Nao   

   Saude_fisica  Saude_mental Dificuldade_caminhar Atividade_fisica  \
0             3            30                  Nao              Sim   
1             0             0                  Nao              Sim   
2            20            30                  Nao              Sim   

  Saude_geral  Horas_sono                Problemas_saude  \
0   Muito boa           5  [Diabetes, Asma, Cancer_pele]   
1   Muito boa           7                          [AVC]   
2   Muito boa           8               [diabetes, Asma]   

                                    Pesquisa   Ano  
0  Principais Indicadores de Doenca Cardiaca  2020  
1  Principais Indicadores de Doenca Cardiaca  2020  
2  Principais Indicadores de Doenca Cardiaca  2020

#HTLM_READER

In [ ]:
import pandas as pd

In [ ]:
dados_html = pd.read_html('/content/filmes_wikipedia.html')

In [ ]:
dados_html

[       0                      1
 0   1998             100 Movies
 1   1999              100 Stars
 2   2000             100 Laughs
 3   2001            100 Thrills
 4   2002           100 Passions
 5   2003  100 Heroes & Villains
 6   2004              100 Songs
 7   2005       100 Movie Quotes
 8   2005              25 Scores
 9   2006             100 Cheers
 10  2006            25 Musicals
 11  2007   100 Movies (Updated)
 12  2008        AFI's 10 Top 10
 13   vte                    vte,
                             Film  Release year              Director  \
 0                   Citizen Kane          1941          Orson Welles   
 1                     Casablanca          1942        Michael Curtiz   
 2                  The Godfather          1972  Francis Ford Coppola   
 3             Gone with the Wind          1939        Victor Fleming   
 4             Lawrence of Arabia          1962            David Lean   
 ..                           ...           ...                   

In [ ]:
type(dados_html)

list

In [ ]:
len(dados_html)

3

In [ ]:
top_filmes = dados_html[1]

In [ ]:
top_filmes

Film  Release year              Director  \
0                   Citizen Kane          1941          Orson Welles   
1                     Casablanca          1942        Michael Curtiz   
2                  The Godfather          1972  Francis Ford Coppola   
3             Gone with the Wind          1939        Victor Fleming   
4             Lawrence of Arabia          1962            David Lean   
..                           ...           ...                   ...   
95                 The Searchers          1956             John Ford   
96              Bringing Up Baby          1938          Howard Hawks   
97                    Unforgiven          1992        Clint Eastwood   
98  Guess Who's Coming to Dinner          1967        Stanley Kramer   
99           Yankee Doodle Dandy          1942        Michael Curtiz   

                      Production companies  Rank  
0                       RKO Radio Pictures     1  
1                    Warner Bros. Pictures     2  
2   Paramount Pictures, Alfran Productions     3  
3          Selznick International Pictures     4  
4                         Horizon Pictures     5  
..                                     ...   ...  
95                  C. V. Whitney Pictures    96  
96                      RKO Radio Pictures    97  
97                     The Malpaso Company    98  
98                       Columbia Pictures    99  
99                   Warner Bros. Pictures   100  

[100 rows x 5 columns]

Editando html é atraves de um csv, foi a maneira que ela abordou, ou seja, precisamos transformar esse dataframe top filmes em um csv. Para uma pessoa mexeer nele no excel, por exemplo.

In [ ]:
top_filmes.to_html('top_filmes.html') #so para vizualização web
top_filmes.to_csv('top_filmes_1998.csv', sep=',',index=False)

#XML

In [ ]:
import pandas as pd

In [ ]:
dados_imdb= pd.read_xml('/content/imdb_top_1000.xml')

In [ ]:
dados_imdb

index                                        Poster_Link  \
0        0  https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1        1  https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2        2  https://m.media-amazon.com/images/M/MV5BMTMxNT...   
3        3  https://m.media-amazon.com/images/M/MV5BMWMwMG...   
4        4  https://m.media-amazon.com/images/M/MV5BMWU4N2...   
..     ...                                                ...   
995    995  https://m.media-amazon.com/images/M/MV5BNGEwMT...   
996    996  https://m.media-amazon.com/images/M/MV5BODk3Yj...   
997    997  https://m.media-amazon.com/images/M/MV5BM2U3Yz...   
998    998  https://m.media-amazon.com/images/M/MV5BZTBmMj...   
999    999  https://m.media-amazon.com/images/M/MV5BMTY5OD...   

                 Series_Title Released_Year Certificate  Runtime  \
0    The Shawshank Redemption          1994           A  142 min   
1               The Godfather          1972           A  175 min   
2             The Dark Knight          2008          UA  152 min   
3      The Godfather: Part II          1974           A  202 min   
4                12 Angry Men          1957           U   96 min   
..                        ...           ...         ...      ...   
995    Breakfast at Tiffany's          1961           A  115 min   
996                     Giant          1956           G  201 min   
997     From Here to Eternity          1953      Passed  118 min   
998                  Lifeboat          1944        None   97 min   
999              The 39 Steps          1935        None   86 min   

                        Genre  IMDB_Rating  \
0                       Drama          9.3   
1                Crime, Drama          9.2   
2        Action, Crime, Drama          9.0   
3                Crime, Drama          9.0   
4                Crime, Drama          9.0   
..                        ...          ...   
995    Comedy, Drama, Romance          7.6   
996            Drama, Western          7.6   
997       Drama, Romance, War          7.6   
998                Drama, War          7.6   
999  Crime, Mystery, Thriller          7.6   

                                              Overview  Meta_score  \
0    Two imprisoned men bond over a number of years...        80.0   
1    An organized crime dynasty's aging patriarch t...       100.0   
2    When the menace known as the Joker wreaks havo...        84.0   
3    The early life and career of Vito Corleone in ...        90.0   
4    A jury holdout attempts to prevent a miscarria...        96.0   
..                                                 ...         ...   
995  A young New York socialite becomes interested ...        76.0   
996  Sprawling epic covering the life of a Texas ca...        84.0   
997  In Hawaii in 1941, a private is cruelly punish...        85.0   
998  Several survivors of a torpedoed merchant ship...        78.0   
999  A man in London tries to help a counter-espion...        93.0   

                 Director              Star1              Star2  \
0          Frank Darabont        Tim Robbins     Morgan Freeman   
1    Francis Ford Coppola      Marlon Brando          Al Pacino   
2       Christopher Nolan     Christian Bale       Heath Ledger   
3    Francis Ford Coppola          Al Pacino     Robert De Niro   
4            Sidney Lumet        Henry Fonda        Lee J. Cobb   
..                    ...                ...                ...   
995         Blake Edwards     Audrey Hepburn     George Peppard   
996        George Stevens   Elizabeth Taylor        Rock Hudson   
997        Fred Zinnemann     Burt Lancaster   Montgomery Clift   
998      Alfred Hitchcock  Tallulah Bankhead        John Hodiak   
999      Alfred Hitchcock       Robert Donat  Madeleine Carroll   

              Star3           Star4  No_of_Votes        Gross  
0        Bob Gunton  William Sadler      2343110   28,341,469  
1        James Caan    Diane Keaton      1620367  134,966,411  
2     Aaron Eckh

In [ ]:
dados_imdb.to_xml('filmes_imbd.xml')

#Trabalhando com banco de dados

In [ ]:
import sqlalchemy

In [ ]:
from sqlalchemy import create_engine, MetaData, table,inspect

In [ ]:
engine = create_engine('sqlite:///: memory :')

In [ ]:
url = 'https://raw.githubusercontent.com/alura-cursos/Pandas/refs/heads/main/clientes_banco.csv'

In [ ]:
import pandas as pd

In [ ]:
dados = pd.read_csv(url)

In [ ]:
dados.head()

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0     5008804     32   Ensino superior  União-estável                2   
1     5008805     32   Ensino superior  União-estável                2   
2     5008806     58      Ensino médio         Casado                2   
3     5008808     52      Ensino médio       Solteiro                1   
4     5008809     52      Ensino médio       Solteiro                1   

    Categoria_de_renda   Ocupacao  Anos_empregado  Rendimento_anual  \
0            Empregado      Outro              12          427500.0   
1            Empregado      Outro              12          427500.0   
2            Empregado  Segurança               3          112500.0   
3  Associado comercial     Vendas               8          270000.0   
4  Associado comercial     Vendas               8          270000.0   

   Tem_carro                   Moradia  
0          1       Apartamento alugado  
1          1       Apartamento alugado  
2          1  Casa/apartamento próprio  
3          0  Casa/apartamento próprio  
4          0  Casa/apartamento próprio

In [ ]:
dados.to_sql('Clientes',engine,index=False)

438463

In [ ]:
inspector = inspect(engine)

In [ ]:
print(inspector.get_table_names())

['Clientes']


query

In [ ]:
query = 'SELECT * FROM Clientes WHERE Categoria_de_renda = "Empregado"'

In [ ]:
empregados =  pd.read_sql(query,engine)
empregados.head()

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0     5008804     32   Ensino superior  União-estável                2   
1     5008805     32   Ensino superior  União-estável                2   
2     5008806     58      Ensino médio         Casado                2   
3     5008815     46   Ensino superior         Casado                2   
4     5112956     46   Ensino superior         Casado                2   

  Categoria_de_renda       Ocupacao  Anos_empregado  Rendimento_anual  \
0          Empregado          Outro              12          427500.0   
1          Empregado          Outro              12          427500.0   
2          Empregado      Segurança               3          112500.0   
3          Empregado  Contabilidade               2          270000.0   
4          Empregado  Contabilidade               2          270000.0   

   Tem_carro                   Moradia  
0          1       Apartamento alugado  
1          1       Apartamento alugado  
2          1  Casa/apartamento próprio  
3          1  Casa/apartamento próprio  
4          1  Casa/apartamento próprio

In [ ]:
empregados.to_sql('empregados',con=engine,index=False)

226059

In [ ]:
pd.read_sql_table('empregados',engine, columns= ['ID_Cliente','Grau_escolaridade','Rendimento_anual'])

ID_Cliente Grau_escolaridade  Rendimento_anual
0          5008804   Ensino superior          427500.0
1          5008805   Ensino superior          427500.0
2          5008806      Ensino médio          112500.0
3          5008815   Ensino superior          270000.0
4          5112956   Ensino superior          270000.0
...            ...               ...               ...
226054     6837905      Ensino médio          355050.0
226055     6837906      Ensino médio          355050.0
226056     6839936      Ensino médio          135000.0
226057     6840222      Ensino médio          103500.0
226058     6842885      Ensino médio          121500.0

[226059 rows x 3 columns]

In [ ]:
from sqlalchemy import text

In [ ]:
query = 'DELETE FROM clientes WHERE ID_Cliente=5008804'
with engine.connect() as conn:
    result = conn.execute(text(query))
    conn.commit()

Atualização da escolaridade do fulano

In [ ]:
query = 'UPDATE clientes SET Grau_escolaridade="Ensino superior" WHERE ID_Cliente=5008808'
with engine.connect() as conn:
    result = conn.execute(text(query))
    conn.commit()

Atualizar o registro do cliente de ID 6840104 que teve o rendimento anual alterado para 300000.

In [ ]:
query2 = ' UPDATE clientes SET Rendimento_anual	= 300000 WHERE ID_Cliente = 6840104'
with engine.connect() as conn2:
  result2 = conn2.execute(text(query2))
  conn2.commit()


In [ ]:
empregados_2 = pd.read_sql(query2,engine)
empregados_2

ResourceClosedError: This result object does not return rows. It has been closed automatically.

Excluir o registro do cliente de ID 5008809, pois essa pessoa não possui mais conta na instituição financeira.


Criar um novo registro de cliente seguindo as especificações abaixo:
ID_Cliente: 6850985
Idade: 33
Grau_escolaridade: Doutorado
Estado_civil: Solteiro
Tamanho_familia: 1
Categoria_de_renda: Empregado
Ocupacao: TI
Anos_empregado: 2
Rendimento_anual: 290000
Tem_carro: 0
Moradia: Casa/apartamento próprio

#Desafios

In [ ]:
import chardet

Verificar a codificação do arquivo

In [ ]:
with open('/content/dados_sus.csv', 'rb') as file:
  print(chardet.detect(file.read()))

FileNotFoundError: [Errno 2] No such file or directory: '/content/dados_sus.csv'

Verifique se o arquivo CSV está separado por vírgula ou ponto e vírgula.


In [ ]:
url_des = '/content/dados_sus.csv'
dados_sus = pd.read_csv(url_des,encoding='ISO-8859-1',sep=';', skiprows=3,skipfooter=9, engine='python')
dados_sus

Unidade da Federação      2008/Jan      2008/Fev      2008/Mar  \
0              Rondônia    1388528,39    2931283,42    1541682,52   
1                  Acre     902416,00    1497206,26    1794028,48   
2              Amazonas    4735529,42    7118990,57    8196635,49   
3               Roraima     657889,53     777939,31     718688,03   
4                  Pará   18864744,11   19553758,20   21937342,70   
5                 Amapá    1120973,34    1268740,33     974651,24   
6             Tocantins    3441592,69    4985688,85    6205612,60   
7              Maranhão   10723506,39   13559679,41   13985142,36   
8                 Piauí    7985142,43   10079928,25   10298372,45   
9                 Ceará   15539821,90   25735101,55   28724625,75   
10  Rio Grande do Norte    5947814,08    9509755,83   10352238,59   
11              Paraíba   10723002,45   11386274,47   12053643,87   
12           Pernambuco   14393949,89   24181400,95   29991302,71   
13              Alagoas    6243726,32    8376622,16    8408336,22   
14              Sergipe    2780750,11    3676483,16    5051473,61   
15                Bahia   21211522,47   28189304,95   34181483,52   
16         Minas Gerais   45671659,76   70194037,10   79231336,95   
17       Espírito Santo    5456723,98    9102212,86   10249595,97   
18       Rio de Janeiro   23025258,52   28394310,93   44567062,94   
19            São Paulo   91678473,48  148677526,82  156087269,23   
20               Paraná   38266347,46   55384670,52   53801318,85   
21       Santa Catarina   16504913,45   25651429,11   26743524,89   
22    Rio Grande do Sul   30931325,46   57363420,01   51547794,07   
23   Mato Grosso do Sul    5377392,11    7002472,61    8423933,19   
24          Mato Grosso    4795299,01    8684882,23    7886243,14   
25                Goiás   15310637,29   22258668,68   21442515,75   
26     Distrito Federal    4745919,19    7632480,67    9647203,26   
27                Total  408424859,23  613174269,21  664043058,38   

        2008/Abr      2008/Mai      2008/Jun      2008/Jul      2008/Ago  \
0     1525314,96    1645953,84    1406150,68    3065279,01    3231494,61   
1     1730469,42    1819443,92    1828496,00    2511754,59    2089107,14   
2     8259378,42    7831399,90    8477273,62    9368858,72    9352532,70   
3      839994,39     862347,96     832440,66     996693,09     894271,18   
4    20842829,69   23249952,88   23240687,56   24002223,56   23341218,03   
5     1141137,32     866346,18    1440978,95    1426659,48    1475551,38   
6     4984482,09    5511002,46    4927934,22    5445602,28    5413710,38   
7    14809416,46   14505727,38   14908091,32   15784899,40   15195312,89   
8    10775976,33   10398360,88   10528253,88   11040107,92   10549176,40   
9    30676457,06   30604453,47   31272395,18   30925391,36   27851438,72   
10   10855108,50   11160168,19   10475071,38   11386349,96   10640817,47   
11   12653528,76   12215107,11   12256928,49   12755413,77   12872578,37   
12   34806197,97   32147987,35   32239094,89   32685640,76   35457038,66   
13    9090794,82    9338771,45   10068210,08    9887061,61    9373432,00   
14    5046063,36    4741151,56    5401259,63    5521904,44    5238398,50   
15   39026934,13   42949566,82   47025972,04   47481695,17   46931248,28   
16   77136537,58   80554440,63   94758522,83   85360064,11   84722528,10   
17   11253833,56   10198619,17   13364258,29   13035374,08   12588595,77   
18   44054604,82   44911021,15   48056212,24   49872379,35   50912706,60   
19  161227125,43  187119397,64  187291126,07  194398399,48  184854516,35   
20   52460306,69   54419620,71   57811688,48   58094241,83   56982774,90   
21   27181488,26   26845308,18   29909762,04   30789135,84   31250854,33   
22   56673338,49   48397733,07   52413429,62   56636376,98   56113054,70   
23    8797487,68    9770421,80   10905392,18   10249299,82    9671553,45   
24    8794109,79    9231645,40   10221763,77    9163314,28    9243148,81   
25   20921212,45   216159

desafio_sheets-google

In [ ]:
https://docs.google.com/spreadsheets/d/1pvBoLyX8kP0TjtUbadVMGdTl4yzm6bHMThhPiqCVtpw/edit?usp=sharing

In [ ]:
sheet_id_des='1pvBoLyX8kP0TjtUbadVMGdTl4yzm6bHMThhPiqCVtpw'
url_des_google=f'https://docs.google.com/spreadsheets/d/{sheet_id_des}/gviz/tq?tqx=out:csv&sheet'
dados_co2_desafio = pd.read_csv(url_des_google)


In [ ]:
dados_co2_desafio.to_csv('dados_desafio_Co2.csv',index=False)

Json_desafio

requests - acessar api's

In [ ]:
import json
import requests

In [ ]:
dados_usuarios = requests.get('https://jsonplaceholder.typicode.com/users')
resultado = json.loads(dados_usuarios.text)

ele so ler com .text

In [ ]:
dados_usuarios_normalizados = pd.json_normalize(resultado,sep='_')

In [ ]:
dados_usuarios_normalizados

id                      name          username                      email  \
0   1             Leanne Graham              Bret          Sincere@april.biz   
1   2              Ervin Howell         Antonette          Shanna@melissa.tv   
2   3          Clementine Bauch          Samantha         Nathan@yesenia.net   
3   4          Patricia Lebsack          Karianne  Julianne.OConner@kory.org   
4   5          Chelsey Dietrich            Kamren   Lucio_Hettinger@annie.ca   
5   6      Mrs. Dennis Schulist  Leopoldo_Corkery    Karley_Dach@jasper.info   
6   7           Kurtis Weissnat      Elwyn.Skiles     Telly.Hoeger@billy.biz   
7   8  Nicholas Runolfsdottir V     Maxime_Nienow       Sherwood@rosamond.me   
8   9           Glenna Reichert          Delphine    Chaim_McDermott@dana.io   
9  10        Clementina DuBuque    Moriah.Stanton     Rey.Padberg@karina.biz   

                   phone        website     address_street address_suite  \
0  1-770-736-8031 x56442  hildegard.org        Kulas Light      Apt. 556   
1    010-692-6593 x09125  anastasia.net      Victor Plains     Suite 879   
2         1-463-123-4447    ramiro.info  Douglas Extension     Suite 847   
3      493-170-9623 x156       kale.biz        Hoeger Mall      Apt. 692   
4          (254)954-1289   demarco.info       Skiles Walks     Suite 351   
5   1-477-935-8478 x6430        ola.org  Norberto Crossing      Apt. 950   
6           210.067.6132       elvis.io          Rex Trail     Suite 280   
7      586.493.6943 x140   jacynthe.com   Ellsworth Summit     Suite 729   
8   (775)976-6794 x41206     conrad.com         Dayna Park     Suite 449   
9           024-648-3804    ambrose.net    Kattie Turnpike     Suite 198   

     address_city address_zipcode address_geo_lat address_geo_lng  \
0     Gwenborough      92998-3874        -37.3159         81.1496   
1     Wisokyburgh      90566-7771        -43.9509        -34.4618   
2   McKenziehaven      59590-4157        -68.6102        -47.0653   
3     South Elvis      53919-4257         29.4572       -164.2990   
4      Roscoeview           33263        -31.8129         62.5342   
5   South Christy      23505-1337        -71.4197         71.7478   
6       Howemouth      58804-1099         24.8918         21.8984   
7       Aliyaview           45169        -14.3990       -120.7677   
8  Bartholomebury      76495-3109         24.6463       -168.8889   
9     Lebsackbury      31428-2261        -38.2386         57.2232   

         company_name                       company_catchPhrase  \
0     Romaguera-Crona    Multi-layered client-server neural-net   
1        Deckow-Crist            Proactive didactic contingency   
2  Romaguera-Jacobson         Face to face bifurcated interface   
3       Robel-Corkery  Multi-tiered zero tolerance productivity   
4         Keebler LLC      User-centric fault-tolerant solution   
5   Considine-Lockman        Synchronised bottom-line interface   
6         Johns Group        Configurable multimedia task-force   
7     Abernathy Group             Implemented secondary concept   
8       Yost and Sons     Switchable contextually-based project   
9          Hoeger LLC         Centralized empowering task-force   

                             company_bs  
0           harness real-time e-markets  
1      synergize scalable supply-chains  
2       e-enable strategic applications  
3  transition cutting-edge web services  
4      revolutionize end-to-end systems  
5      e-enable innovative applications  
6         generate enterprise e-tailers  
7         e-enable extensible e-tailers  
8      aggregate real-time technologies  
9              target end-to-end models

Desafio HTML

In [ ]:
dados_wiki = pd.read_html('/content/Lista de países por população.html')

In [ ]:
len(dados_wiki)

3

In [ ]:
planilha_wiki=dados_wiki[0]

Unnamed: 0 Posição    País (ou território dependente) Estimativa da ONU  \
0           NaN       1                              Índia     1 417 492 000   
1           NaN       2                      China[nota 1]     1.407.934.000   
2           NaN       3                     Estados Unidos       342.181.000   
3           NaN       4                          Indonésia       285.783.000   
4           NaN       5                          Paquistão       256 204 000   
..          ...     ...                                ...               ...   
245         NaN       –     Ilha de Ascensão (Reino Unido)             1 100   
246         NaN     195                           Vaticano               879   
247         NaN       –  Ilhas Cocos (Keeling) (Austrália)               605   
248         NaN       –     Tristão da Cunha (Reino Unido)               264   
249         NaN       –       Ilhas Pitcairn (Reino Unido)                67   

    Crescimento desde a última estimativa      Estimativa Oficial  
0                             -12 692 000      Estimativa oficial  
1                              -1 390 000           Censo oficial  
2                              -2 084 000           Censo oficial  
3                              +3 337 000      Estimativa oficial  
4                              +6 567 000      Estimativa oficial  
..                                    ...                     ...  
245                                   NaN  Estimativa oficial[10]  
246                                   NaN      Estimativa oficial  
247                                   NaN      Estimativa oficial  
248                                   NaN  Estimativa oficial[11]  
249                                   NaN      Estimativa oficial  

[250 rows x 6 columns]

In [ ]:
dados_paises = planilha_wiki[['País (ou território dependente)','Estimativa da ONU']]

In [ ]:
dados_paises

País (ou território dependente) Estimativa da ONU
0                                Índia     1 417 492 000
1                        China[nota 1]     1.407.934.000
2                       Estados Unidos       342.181.000
3                            Indonésia       285.783.000
4                            Paquistão       256 204 000
..                                 ...               ...
245     Ilha de Ascensão (Reino Unido)             1 100
246                           Vaticano               879
247  Ilhas Cocos (Keeling) (Austrália)               605
248     Tristão da Cunha (Reino Unido)               264
249       Ilhas Pitcairn (Reino Unido)                67

[250 rows x 2 columns]

In [ ]:
dados_paises.to_html('dados_paises.html', index= False)

In [ ]:
dados_paises.to_csv('dados_paises.csv', index=False)

In [ ]:
pd.read_csv('/content/dados_paises.csv')

País (ou território dependente) Estimativa da ONU
0                                Índia     1 417 492 000
1                        China[nota 1]     1.407.934.000
2                       Estados Unidos       342.181.000
3                            Indonésia       285.783.000
4                            Paquistão       256 204 000
..                                 ...               ...
245     Ilha de Ascensão (Reino Unido)             1 100
246                           Vaticano               879
247  Ilhas Cocos (Keeling) (Austrália)               605
248     Tristão da Cunha (Reino Unido)               264
249       Ilhas Pitcairn (Reino Unido)                67

[250 rows x 2 columns]

In [ ]:
import pandas as pd

In [ ]:
url_desafio = 'https://raw.githubusercontent.com/alura-cursos/Pandas/main/clientes_banco.csv'

In [ ]:
import sqlalchemy

In [ ]:
from sqlalchemy import create_engine,MetaData, table,inspect, text

In [ ]:
engine=create_engine('sqlite:///: memory:')

In [ ]:
dados_cvs_clientes = pd.read_csv(url_desafio)

In [ ]:
dados_cvs_clientes

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0          5008804     32   Ensino superior  União-estável                2   
1          5008805     32   Ensino superior  União-estável                2   
2          5008806     58      Ensino médio         Casado                2   
3          5008808     52      Ensino médio       Solteiro                1   
4          5008809     52      Ensino médio       Solteiro                1   
...            ...    ...               ...            ...              ...   
438458     6840104     62      Ensino médio     Divorciado                1   
438459     6840222     43      Ensino médio       Solteiro                1   
438460     6841878     22   Ensino superior       Solteiro                1   
438461     6842765     59      Ensino médio         Casado                2   
438462     6842885     51      Ensino médio         Casado                2   

         Categoria_de_renda          Ocupacao  Anos_empregado  \
0                 Empregado             Outro              12   
1                 Empregado             Outro              12   
2                 Empregado         Segurança               3   
3       Associado comercial            Vendas               8   
4       Associado comercial            Vendas               8   
...                     ...               ...             ...   
438458          Pensionista             Outro               0   
438459            Empregado  Construção Civil               8   
438460  Associado comercial            Vendas               1   
438461          Pensionista             Outro               0   
438462            Empregado            Vendas               3   

        Rendimento_anual  Tem_carro                   Moradia  
0               427500.0          1       Apartamento alugado  
1               427500.0          1       Apartamento alugado  
2               112500.0          1  Casa/apartamento próprio  
3               270000.0          0  Casa/apartamento próprio  
4               270000.0          0  Casa/apartamento próprio  
...                  ...        ...                       ...  
438458          135000.0          0  Casa/apartamento próprio  
438459          103500.0          0  Casa/apartamento próprio  
438460           54000.0          0          Mora com os pais  
438461           72000.0          0  Casa/apartamento próprio  
438462          121500.0          0  Casa/apartamento próprio  

[438463 rows x 11 columns]

Perfeito, agora que voce leu os arquivos, vamos transformar ele em um banco de dados sql

In [ ]:
dados_clientes_sql = dados_cvs_clientes.to_sql('costumers',engine,index = False)

ValueError: Table 'costumers' already exists.

In [ ]:
inspector = inspect(engine)

se vc quer obter agora informação sobre a tabela vc tem que operar com a engine

In [ ]:
print(inspector.get_table_names())

['Clientes', 'costumers']


In [ ]:
dados_clientes_sql

438463

Atualizar o registro do cliente de ID 6840104 que teve o rendimento anual alterado para 300000.

In [ ]:
query_1_desafio = "UPDATE costumers SET Rendimento_anual = 300000 WHERE ID_Cliente = 6840104"

applying the query

In [ ]:
with engine.connect() as conn:
  result =conn.execute(text(query_1_desafio))
  conn.commit()

In [ ]:
dados_cvs_clientes

ID_Cliente  Idade Grau_escolaridade   Estado_civil  Tamanho_familia  \
0          5008804     32   Ensino superior  União-estável                2   
1          5008805     32   Ensino superior  União-estável                2   
2          5008806     58      Ensino médio         Casado                2   
3          5008808     52      Ensino médio       Solteiro                1   
4          5008809     52      Ensino médio       Solteiro                1   
...            ...    ...               ...            ...              ...   
438458     6840104     62      Ensino médio     Divorciado                1   
438459     6840222     43      Ensino médio       Solteiro                1   
438460     6841878     22   Ensino superior       Solteiro                1   
438461     6842765     59      Ensino médio         Casado                2   
438462     6842885     51      Ensino médio         Casado                2   

         Categoria_de_renda          Ocupacao  Anos_empregado  \
0                 Empregado             Outro              12   
1                 Empregado             Outro              12   
2                 Empregado         Segurança               3   
3       Associado comercial            Vendas               8   
4       Associado comercial            Vendas               8   
...                     ...               ...             ...   
438458          Pensionista             Outro               0   
438459            Empregado  Construção Civil               8   
438460  Associado comercial            Vendas               1   
438461          Pensionista             Outro               0   
438462            Empregado            Vendas               3   

        Rendimento_anual  Tem_carro                   Moradia  
0               427500.0          1       Apartamento alugado  
1               427500.0          1       Apartamento alugado  
2               112500.0          1  Casa/apartamento próprio  
3               270000.0          0  Casa/apartamento próprio  
4               270000.0          0  Casa/apartamento próprio  
...                  ...        ...                       ...  
438458          135000.0          0  Casa/apartamento próprio  
438459          103500.0          0  Casa/apartamento próprio  
438460           54000.0          0          Mora com os pais  
438461           72000.0          0  Casa/apartamento próprio  
438462          121500.0          0  Casa/apartamento próprio  

[438463 rows x 11 columns]

Excluir o registro do cliente de ID 5008809, pois essa pessoa não possui mais conta na instituição financeira.

Criar um novo registro de cliente seguindo as especificações abaixo:
ID_Cliente: 6850985
Idade: 33
Grau_escolaridade: Doutorado
Estado_civil: Solteiro
Tamanho_familia: 1
Categoria_de_renda: Empregado
Ocupacao: TI
Anos_empregado: 2
Rendimento_anual: 290000
Tem_carro: 0
Moradia: Casa/apartamento próprio